src/combinatorics/composition.lean

```diff
@@ -0,0 +1,584 @@
+/-
+Ccompoyright (c) 2020 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import data.fintype.card tactic.omega tactic.tidy
+
+/-!
+# Compositions
+
+A composition of an integer `n` is a decomposition of `{0, ..., n-1}` into blocks of consecutive
+integers. Equivalently, it is a decomposition `n = i₀ + ... + i_{k-1}` into a sum of positive
+integers, where the `iⱼ` are the lengths of the blocks. This notion is closely related to that
+of a partition of `n`, but in a composition of `n` the order of the `iⱼ`s matters.
+
+We represent a composition of `n` as a subset of `{0, ..., n}` containing `0` and `n`, where
+the elements of the subset (but `n`) correspond to the leftmost points of each block. Another
+implementation would be a subset of `{0, ..., n-1}` containing `0`, but this fails in the edge case
+`n = 0`. Another implementation would be as an integer `k` and a function `fin k → ℕ` summing to
+`n`, but this turns out to be less handy because of dependent type issues.
+
+## Main functions
+
+* `compo : composition n` is a structure, made of a subset of `{0, ..., n}` containing `0` and
+  `n`, representing a composition of the natural number `n`.
```

Comment:
I'm not sure why, but I had trouble parsing this on the first read --- I guess I didn't realise `comp` was intended a variable name. Would renaming it just `c` be okay?

Reply:
Rofl. Bad search and replace...

Reply:
I renamed `compo` to `c` everywhere in the file, and clarified the sentence.